src/dbnode/storage/mediator.go

```diff
@@ -226,3 +246,88 @@ func (m *mediator) reportLoop() {
+// concurrently with an inconsistent view of time. Each time the filesystem processes want to run they first
+// register for the next barrier by calling wait(). Once a tick completes it will detect that the filesystem
+// processes are waiting for the next barrier by calling hasWaiter() at which point it will update the mediator
+// time and propagate that information to the filesystem processes via a call to release(). If the filesystem
+// processes are still running when the tick completes, it will just tick again but use the same startTime as
+// the previous run.
+//
+// This cooperation ensures that multiple ticks can run during a single run of filesystem processes (although
+// each tick will run with the same startTime), but that if a tick and run of filesystem processes are executing
+// concurrently they will always have the same value for startTime.
+//
+// One implication of this scheme is that once a run of filesystem processes completes it will always have to wait
+// until the currently executing tick completes before performing the next run, but in practice this should not be
+// much of an issue.
+//
+//  ____________       ___________
+// | Flush (t0) |     | Tick (t0) |
+// |            |     |           |
+// |            |     |___________|
+// |            |      ___________
+// |            |     | Tick (t0) |
+// |            |     |           |
+// |            |     |___________|
+// |            |      ___________
+// |____________|     | Tick (t0) |
+//  barrier.wait()    |           |
+//                    |___________|
+//                    mediatorTime = t1
+//                    barrier.release()
+// -------------------------------------
+//  ____________       ___________
+// | Flush (t1) |     | Tick (t1) |
+// |            |     |           |
+// |            |     |___________|
+// |            |      ___________
+// |            |     | Tick (t1) |
+// |            |     |           |
+// |            |     |___________|
+// |            |      ___________
+// |____________|     | Tick (t1) |
+//  barrier.wait()    |           |
+//                    |___________|
+//                    barrier.release()
+// ------------------------------------
+type mediatorTimeBarrier struct {
+	sync.Mutex
+	isWaiting bool
+	doneCh    chan time.Time
+}
+
+func (b *mediatorTimeBarrier) wait() (time.Time, error) {
+	b.Lock()
+	if b.isWaiting {
+		b.Unlock()
+		return time.Time{}, errMediatorTimeBarrierAlreadyWaiting
+	}
+	b.isWaiting = true
+	b.Unlock()
+
+	t := <-b.doneCh
+
+	b.Lock()
+	b.isWaiting = false
+	b.Unlock()
+	return t, nil
+}
+
+func (b *mediatorTimeBarrier) release(t time.Time) {
+	b.doneCh <- t
+}
+
+func (b *mediatorTimeBarrier) hasWaiter() bool {
+	b.Lock()
+	isWaiting := b.isWaiting
+	b.Unlock()
+	return isWaiting
+}
+
+func newMediatorTimeBarrier() mediatorTimeBarrier {
+	return mediatorTimeBarrier{
+		doneCh: make(chan time.Time, 0),
```

Comment:
I kind of wonder if the `isWaiting` thing is really necessary.

Couldn't we just have two channels, the current `doneCh` with size zero (maybe renamed `waitCh`, and an exclusive `exclusiveTokenCh` with a pre-seeded token  and then the release code is basically `releaseIfWaiting()` (this will reduce the if/else in the actual loop):
```go
func (b *mediatorTimeBarrier) releaseIfWaiting(t time.Time) {
  select {
  case b.waitCh <- t:
  default:
    // not waiting
  }
}

func (b *mediatorTimeBarrier) exclusiveWait() (time.Time, error) {
  var token struct{}
  select {
  case token = <- b.exclusiveTokenCh:
  default: 
    return time.Time{}, errMediatorTimeBarrierAlreadyWaiting
  }
  
  t := <- b.waitCh
  
  b.exclusiveCh <- token
  return t, nil
}
```

It reduces the other loop to just:
```go
			newMediatorTime := m.nowFn()
			if newMediatorTime.Before(mediatorTime) {
				// .. error
				continue
			}

			mediatorTime = newMediatorTime

			m.mediatorTimeBarrier.releaseIfWaiting()

			if err := m.Tick(force, mediatorTime); err != nil {
				log.Error("error within tick", zap.Error(err))
			}
```

Reply:
refactored the whole thing, let me know what you think